arangod/Aql/AqlValue.cpp

```diff
@@ -1339,64 +1068,52 @@ int AqlValue::Compare(velocypack::Options const* options, AqlValue const& left,
         return 1;
       }
       return 0;
-    }
   }
   return 0;
 }
 
-AqlValue::AqlValue() noexcept {
-  // construct a slice of type None
-  // we will simply zero-initialize the two 64 bit words
-  _data.words[0] = 0;
-  _data.words[1] = 0;
-
-  // VPACK_INLINE must have a value of 0, and VPackSlice::None must be equal
-  // to a NUL byte too
-  static_assert(AqlValueType::VPACK_INLINE == 0,
-                "invalid value for VPACK_INLINE");
-}
+AqlValue::AqlValue() noexcept { erase(); }
 
-AqlValue::AqlValue(std::unique_ptr<uint8_t[]> data) noexcept {
+AqlValue::AqlValue(std::unique_ptr<std::string>& data) noexcept {
   TRI_ASSERT(data);
-  velocypack::Slice slice{data.get()};
-  setManagedSliceData(MemoryOriginType::New, slice.byteSize());
-  _data.managedSliceMeta.managedPointer = data.release();
+  auto size = data->size();
+  TRI_ASSERT(size >= 1);
+  VPackSlice slice{reinterpret_cast<uint8_t const*>(data->data())};
+  TRI_ASSERT(size == slice.byteSize());
+  TRI_ASSERT(!slice.isExternal());
+  if (size < sizeof(AqlValue)) {
+    initFromSlice(slice, size);
+  } else {
+    setType(AqlValueType::VPACK_MANAGED_STRING);
+    _data.managedStringMeta.pointer = data.release();
+  }
 }
 
 AqlValue::AqlValue(uint8_t const* pointer) noexcept {
+  // TODO(MBkkt) Is external possible here?
   // we must get rid of Externals first here, because all
   // methods that use VPACK_SLICE_POINTER expect its contents
   // to be non-Externals
-  if (*pointer == '\x1d') {
-    // an external
-    setPointer<false>(VPackSlice(pointer).resolveExternals().begin());
-  } else {
-    setPointer<false>(pointer);
-  }
-  TRI_ASSERT(!VPackSlice(_data.pointerMeta.pointer).isExternal());
+  setPointer<false>(VPackSlice{pointer}.resolveExternals().begin());
 }
 
-AqlValue::AqlValue(AqlValue const& other, void* data) noexcept {
+AqlValue::AqlValue(AqlValue const& other, void const* data) noexcept {
   TRI_ASSERT(data != nullptr);
-  setType(other.type());
-  switch (other.type()) {
+  auto t = other.type();
+  setType(t);
+  switch (t) {
     case VPACK_MANAGED_SLICE:
       _data.managedSliceMeta.lengthOrigin =
           other._data.managedSliceMeta.lengthOrigin;
-      _data.managedSliceMeta.managedPointer = static_cast<uint8_t*>(data);
+      _data.managedSliceMeta.pointer =
+          const_cast<uint8_t*>(static_cast<uint8_t const*>(data));
       break;
-    case VPACK_SLICE_POINTER:
```

Comment:
Is this case removed intentionally? Why?

Reply:
Because this function should be called only for destructible AqlValue.
So it looks like "typo" in old code

Reply:
ok

Reply:
You can also see assert(false) in default case, so it also will be checked by tests.

Reply:
Same for `AqlValue::data()`